Bottom: 2516cf5fbc3572924f6f2139d5171f735f331563
Top:    8a36300e13a698a72d60efee2dc4e90cb76a6ef3
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-01-03 11:16:05 +0800

Refresh of tuntap-xdp-transmission

---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index 2586cc3..c14de98 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1271,12 +1271,22 @@ static int tun_xdp_xmit(struct net_device *dev, struct xdp_buff *xdp)
 static void tun_xdp_flush(struct net_device *dev)
 {
 	struct tun_struct *tun = netdev_priv(dev);
-	struct tun_file *tfile = tun->tfiles[0];
+	struct tun_file *tfile;
+	u32 numqueues;
+
+	rcu_read_lock();
+
+	numqueues = READ_ONCE(tun->numqueues);
+	BUG_ON(!numqueues);
+	tfile = rcu_dereference(tun->tfiles[smp_processor_id() %
+					    numqueues]);
 
 	/* Notify and wake up reader process */
 	if (tfile->flags & TUN_FASYNC)
 		kill_fasync(&tfile->fasync, SIGIO, POLL_IN);
 	tfile->socket.sk->sk_data_ready(tfile->socket.sk);
+
+	rcu_read_unlock();
 }
 
 static const struct net_device_ops tap_netdev_ops = {
